File: solutions/R/ml_foundations_ex3.R
# =============================================================================
# Chapter 7, Exercise 3: The Bias-Variance Tradeoff in Practice
# Fit polynomials of degree 1, 3, 5, 10, and 20 to a training set.
# Plot fitted curves, compute training/test error, identify optimal degree.
# =============================================================================

library(tidyverse)

# --- Generate data ---
set.seed(42)
x_train <- sort(runif(50, 0, 10))
y_train <- sin(x_train) + rnorm(50, 0, 0.3)
x_test <- sort(runif(200, 0, 10))
y_test <- sin(x_test) + rnorm(200, 0, 0.3)

train_df <- tibble(x = x_train, y = y_train)
test_df  <- tibble(x = x_test,  y = y_test)

# --- Fit polynomials and compute RMSE ---
degrees <- c(1, 3, 5, 10, 20)
results <- tibble(degree = integer(), train_rmse = double(), test_rmse = double())

# Also store predictions for plotting
plot_data <- tibble()

for (d in degrees) {
  # Fit polynomial of degree d
  fit <- lm(y ~ poly(x, degree = d, raw = TRUE), data = train_df)

  # Predictions on train and test
  pred_train <- predict(fit, newdata = train_df)
  pred_test  <- predict(fit, newdata = test_df)

  # Compute RMSE
  rmse_train <- sqrt(mean((y_train - pred_train)^2))
  rmse_test  <- sqrt(mean((y_test - pred_test)^2))

  results <- bind_rows(results,
                       tibble(degree = d, train_rmse = rmse_train, test_rmse = rmse_test))

  # Smooth curve for plotting
  x_grid <- seq(0, 10, length.out = 300)
  pred_grid <- predict(fit, newdata = tibble(x = x_grid))
  # Clip extreme predictions for high-degree polynomials
  pred_grid <- pmin(pmax(pred_grid, -3), 3)
  plot_data <- bind_rows(plot_data,
                         tibble(x = x_grid, y_pred = pred_grid,
                                degree = paste("Degree", d)))
}

# --- Print RMSE results ---
cat("Polynomial Regression: Training vs Test RMSE\n")
cat("=============================================\n")
print(results)

# --- Part 2: Plot fitted curves ---
p1 <- ggplot() +
  geom_point(data = train_df, aes(x, y), alpha = 0.5, size = 2) +
  geom_line(data = plot_data, aes(x, y_pred, color = degree), linewidth = 1) +
  geom_line(data = tibble(x = seq(0, 10, 0.01), y = sin(seq(0, 10, 0.01))),
            aes(x, y), linetype = "dashed", color = "black", linewidth = 0.8) +
  labs(title = "Polynomial Fits of Varying Complexity",
       subtitle = "Dashed line = true function sin(x)",
       x = "x", y = "y", color = "Polynomial") +
  theme_minimal(base_size = 14) +
  theme(legend.position = "top")

print(p1)

# --- Part 3: Plot training vs test error ---
results_long <- results %>%
  pivot_longer(cols = c(train_rmse, test_rmse),
               names_to = "set", values_to = "rmse") %>%
  mutate(set = ifelse(set == "train_rmse", "Training", "Test"))

p2 <- ggplot(results_long, aes(x = degree, y = rmse, color = set)) +
  geom_line(linewidth = 1.2) +
  geom_point(size = 3) +
  labs(title = "Training vs Test RMSE by Polynomial Degree",
       x = "Polynomial Degree", y = "RMSE", color = "Dataset") +
  theme_minimal(base_size = 14) +
  theme(legend.position = "top")

print(p2)

# --- Part 4: Interpretation ---
best_degree <- results$degree[which.min(results$test_rmse)]
cat("\nBest polynomial degree (lowest test RMSE):", best_degree, "\n")
cat("\nInterpretation (Bias-Variance Tradeoff):\n")
cat("- Degree 1 (linear): High bias -- too simple to capture the sine curve.\n")
cat("  Underfits both training and test data.\n")
cat("- Degree 3-5: Good balance. Captures the main curvature of sin(x)\n")
cat("  without fitting noise. Test error is minimized here.\n")
cat("- Degree 10-20: High variance -- the polynomial wiggles to fit\n")
cat("  training noise. Training error drops but test error increases.\n")
cat("  This is classic overfitting.\n")
cat("\nThe optimal degree (~3-5) sits at the sweet spot of the bias-variance\n")
cat("tradeoff, where the model is flexible enough to capture the true\n")
cat("pattern but not so flexible that it memorizes noise.\n")

File: solutions/python/ml_foundations_ex3.py
# =============================================================================
# Chapter 7, Exercise 3: The Bias-Variance Tradeoff in Practice
# Fit polynomials of degree 1, 3, 5, 10, and 20 to a training set.
# Plot fitted curves, compute training/test error, identify optimal degree.
# =============================================================================

import numpy as np
import matplotlib.pyplot as plt
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

# --- Generate data ---
np.random.seed(42)
x_train = np.sort(np.random.uniform(0, 10, 50))
y_train = np.sin(x_train) + np.random.normal(0, 0.3, 50)
x_test = np.sort(np.random.uniform(0, 10, 200))
y_test = np.sin(x_test) + np.random.normal(0, 0.3, 200)

# --- Fit polynomials and compute RMSE ---
degrees = [1, 3, 5, 10, 20]
results = []
predictions = {}

for d in degrees:
    # Create polynomial features and fit
    poly = PolynomialFeatures(degree=d, include_bias=False)
    X_train_poly = poly.fit_transform(x_train.reshape(-1, 1))
    X_test_poly = poly.transform(x_test.reshape(-1, 1))

    model = LinearRegression()
    model.fit(X_train_poly, y_train)

    # Predictions
    pred_train = model.predict(X_train_poly)
    pred_test = model.predict(X_test_poly)

    # RMSE
    rmse_train = np.sqrt(mean_squared_error(y_train, pred_train))
    rmse_test = np.sqrt(mean_squared_error(y_test, pred_test))

    results.append({'degree': d, 'train_rmse': rmse_train, 'test_rmse': rmse_test})

    # Smooth curve for plotting
    x_grid = np.linspace(0, 10, 300).reshape(-1, 1)
    X_grid_poly = poly.transform(x_grid)
    pred_grid = model.predict(X_grid_poly)
    # Clip extreme predictions for high-degree polynomials
    pred_grid = np.clip(pred_grid, -3, 3)
    predictions[d] = (x_grid.ravel(), pred_grid)

# --- Print RMSE results ---
print("Polynomial Regression: Training vs Test RMSE")
print("=" * 50)
print(f"{'Degree':>8s}  {'Train RMSE':>12s}  {'Test RMSE':>12s}")
print("-" * 50)
for r in results:
    print(f"{r['degree']:>8d}  {r['train_rmse']:>12.4f}  {r['test_rmse']:>12.4f}")

# --- Part 2: Plot fitted curves ---
fig, ax = plt.subplots(figsize=(10, 6))
ax.scatter(x_train, y_train, alpha=0.5, s=30, label='Training data', zorder=5)
x_true = np.linspace(0, 10, 300)
ax.plot(x_true, np.sin(x_true), 'k--', linewidth=1.5, label='True function sin(x)')

colors = ['#E69F00', '#56B4E9', '#009E73', '#D55E00', '#CC79A7']
for (d, color) in zip(degrees, colors):
    x_g, pred_g = predictions[d]
    ax.plot(x_g, pred_g, linewidth=1.5, color=color, label=f'Degree {d}')

ax.set_xlabel('x')
ax.set_ylabel('y')
ax.set_title('Polynomial Fits of Varying Complexity')
ax.legend(loc='upper right', fontsize=9)
ax.set_ylim(-3, 3)
plt.tight_layout()
plt.show()

# --- Part 3: Plot training vs test error ---
degs = [r['degree'] for r in results]
train_rmses = [r['train_rmse'] for r in results]
test_rmses = [r['test_rmse'] for r in results]

fig, ax = plt.subplots(figsize=(8, 5))
ax.plot(degs, train_rmses, 'o-', linewidth=2, label='Training RMSE', color='steelblue')
ax.plot(degs, test_rmses, 'o-', linewidth=2, label='Test RMSE', color='darkorange')
ax.set_xlabel('Polynomial Degree')
ax.set_ylabel('RMSE')
ax.set_title('Training vs Test RMSE by Polynomial Degree')
ax.legend()
plt.tight_layout()
plt.show()

# --- Part 4: Interpretation ---
best_idx = np.argmin(test_rmses)
best_degree = degs[best_idx]
print(f"\nBest polynomial degree (lowest test RMSE): {best_degree}")
print("\nInterpretation (Bias-Variance Tradeoff):")
print("- Degree 1 (linear): High bias -- too simple to capture the sine curve.")
print("  Underfits both training and test data.")
print("- Degree 3-5: Good balance. Captures the main curvature of sin(x)")
print("  without fitting noise. Test error is minimized here.")
print("- Degree 10-20: High variance -- the polynomial wiggles to fit")
print("  training noise. Training error drops but test error increases.")
print("  This is classic overfitting.")
print("\nThe optimal degree (~3-5) sits at the sweet spot of the bias-variance")
print("tradeoff, where the model is flexible enough to capture the true")
print("pattern but not so flexible that it memorizes noise.")
